ovc/client.go

```diff
@@ -133,69 +147,101 @@ func NewClient(c *Config) (*Client, error) {
 }
 
 // async adds "async=true" flag to all API calls
-func async(req *http.Request) (*http.Request, error) {
+func (c *Client) async(req *http.Request) ([]byte, error) {
 	// fetch request body to the string
 	jsonMap := make(map[string]interface{})
 
 	if req.Body != nil {
 		reqBody, err := ioutil.ReadAll(req.Body)
 		if err != nil {
+			c.logger.Errorf("Failed to read body from http request: %s", err)
 			return nil, err
 		}
 		err = json.Unmarshal(reqBody, &jsonMap)
 		if err != nil {
+			c.logger.Errorf("req.Method: %s", req.Method)
+			c.logger.Errorf("req.URL: %s", req.URL.String())
+			c.logger.Errorf("Failed to marshal json body into an object: %s\njson body:\n%s", err, string(reqBody))
 			return nil, err
 		}
 	}
 
 	jsonMap["_async"] = true
-	configJSON, err := json.Marshal(jsonMap)
-	if err != nil {
-		return nil, err
-	}
-	newReq, err := http.NewRequest(req.Method, req.URL.String(), bytes.NewBuffer(configJSON))
-	if err != nil {
-		return nil, err
-	}
-	return newReq, nil
+	return json.Marshal(jsonMap)
 }
 
 // Do sends and API Request and returns the body as an array of bytes
-func (c *Client) Do(req *http.Request) ([]byte, error) {
-	req, err := async(req) // make request asynchronous
-	if err != nil {
-		return nil, err
-	}
-	client := &http.Client{}
-	tokenString, err := c.JWT.Get()
-	if err != nil {
-		return nil, err
-	}
-	req.Header.Set("Authorization", fmt.Sprintf("bearer %s", tokenString))
-	req.Header.Set("Content-Type", "application/json")
-	resp, err := client.Do(req)
-	if resp != nil {
-		defer resp.Body.Close()
-	}
+func (c *Client) Do(req *http.Request, timeout ResponseTimeout) ([]byte, error) {
```

Comment:
This changes an interface that's technically public (but *probably* not used) - do we want to offer backward compatibility?

Reply:
I have thought about making this method private as i don't see why it should be public. If we lack support for a specific API call we should add it with a nice typed interface. So i'd rather break the api. The only packages using this library is our own, so i don't think its a big issue to break it. We can tag with a new major version, then its even more ok i guess.